### DISCUSSION TITLE
Cryostat declarative config

### DISCUSSION THREAD

**Discussion by grzesuav** [cryostatio/cryostat#714] [Q&A]
Hi,

is there a way to create i.e. automated recording rule declaratively, like in helm chart config (not via UI) ?

I am looking to harden our setup and this is one of the thing which come first.

**Comment by andrewazores** [comment]
Hi @grzesuav ,

It's an old idea, but not one we've ever gotten around to implementing (yet). I can't find an exact matching issue, but this one is similar: https://github.com/cryostatio/cryostat-operator/issues/464

I still think it would be a really good feature to have. I'm imagining a rough outline like this:

## Cryostat container
1. Add a config property (environment variable) to allow specifying an "import search path" for Automated Rules. By default this will be the empty string indicating no import path. This can be overridden to a path within the local filesystem. On startup, Cryostat will walk the filesystem below the given path looking for any files.
2. Files found in the import path are expected to be Automated Rules JSON files. Cryostat will attempt to deserialize them and create the corresponding Automated Rules.
3. If a Rule already exists or otherwise conflicts, Cryostat should log a warning and continue.

## Helm Chart
1. Add a configurable value for a list of ConfigMap (or Secret?) names. These must be located in the chart installation namespace. If this list is non-empty, then the chart will set the "import search path" config property on the Cryostat container, and mount all of the ConfigMaps to that location as files.
2. The user will be responsible for creating their ConfigMaps in advance before installing Cryostat, and populating them with Automated Rules JSON contents. This will be done separately from the Helm Chart, not as a direct Helm Chart feature.

## Operator
1. Add a CRD field for a list of ConfigMap (or Secret?) names - same as what the Helm Chart does above.

----

Does this sound like it would achieve what you're looking for?

**Reply by grzesuav** [reply]
Yes, it would !

**Reply by andrewazores** [reply]
Cool, I'll try to find time to work on this. There is already precedent for it with [custom event templates](https://github.com/cryostatio/cryostat/blob/25eaba2800b2830e1ac1efebd4ee22b275f7f2ed/src/main/java/io/cryostat/events/S3TemplateService.java#L105) in Cryostat as well as [the Operator](https://github.com/cryostatio/cryostat-operator/blob/main/docs/config.md#custom-event-templates) already, so it should not be too difficult.

**Reply by andrewazores** [reply]
#718
